src/Test/QuickCheck/Generic.purs

```diff
@@ -0,0 +1,64 @@
+-- | Generic deriving for `Arbitrary` and `Coarbitrary` instances.
+
+module Test.QuickCheck.Generic 
+  ( gArbitrary
+  , gCoarbitrary
+  ) where
+
+import Prelude
+
+import Data.Maybe (Maybe(..))
+import Data.Maybe.Unsafe (fromJust)
+import Data.Array (uncons)
+import Data.Traversable (traverse)
+import Data.Generic
+import Data.Monoid.Endo
+import Data.Foldable (Foldable, foldMap)
+
+import Control.Lazy (defer)
+
+import Test.QuickCheck.Gen
+import Test.QuickCheck.Arbitrary
+
+-- | Generate arbitrary values for any `Generic` data structure.
+-- |
+-- | _Note_: this function _is_ partial, since it does not handle ADTs with zero constructors
+-- | such as `Void`.
```

Comment:
Ok, and this fixes the stack failure, but I don't know if it actually makes things better in terms of Void... https://gist.github.com/gbaz/5eb8be9a4603ad59a8fa

Note that defer didn't need to be reintroduced. The trick is just not mapping running over all the constructors before picking one. This revealed the need for a lighter-weight oneOf in Gen, which I inlined here.

Even if the Maybe stuff gets stripped back out, I think this approach is better and less delicate than the explicit laziness...